### PR TITLE
initial implementation of orderElmModuleByDependency & insertModule functions

### PR DESCRIPTION
**Pull request by osborneMs**
…unctions

# Terms

> THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE TERMS OF THE CCLA DATED 2017-11-07 WITH FINOS/LINUX FOUNDATION (FORMERLY THE SYMPHONY SOFTWARE FOUNDATION CCLA).
>
> THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

### CODE REVIEW

File: src/Morphir/IR/Repo.elm [hunk 1]
```diff
@@ -124,9 +126,91 @@ parseNewElmModules fileChanges =
...
+           |> Result.andThen (\functionAResult ->
+               case functionAResult of
+                   Err err ->
+                       case  err of
+                           _ ->
+                               CycleDetected
+                   Ok _ ->
+                       parsedModules
+           )
+    -}
```
Comment (AttilaMihaly) [inline src/Morphir/IR/Repo.elm:164]: Please take this out.

File: src/Morphir/IR/Repo.elm [hunk 2]
```diff
@@ -124,9 +126,91 @@ parseNewElmModules fileChanges =
...
+            let
+                moduleName : ModuleName
+                moduleName =
+                    ParsedModule.moduleName parsedModule
+                        |> toIRModuleName repo.packageName
+
+                moduleDependencies : List ModuleName
+                moduleDependencies =
+                    ParsedModule.importedModules parsedModule
+                        |> List.map (\modName -> toIRModuleName repo.packageName modName)
```
Comment (AttilaMihaly) [inline src/Morphir/IR/Repo.elm:181]: This should check if the Elm module name starts with the current package name and ignore any imports that don't satisfy that.

File: src/Morphir/IR/Repo.elm [hunk 3]
```diff
@@ -124,9 +126,91 @@ parseNewElmModules fileChanges =
...
+                |> List.foldl insertEdge graph
+    in
+    parsedModules
+        |> List.foldl foldFunction (Ok moduleGraph)
+        |> Result.andThen
+            (\finalGraph ->
+                finalGraph
+                    |> DAG.forwardTopologicalOrdering
+                    |> List.concat
+                    |> List.map2
```
Comment (AttilaMihaly) [inline src/Morphir/IR/Repo.elm:208]: You can remove `map2` and change `Result.andThen` to `Result.map`

**Review by AttilaMihaly — CHANGES_REQUESTED** [review]

File: src/Morphir/IR/Repo.elm (hunk 1 above)
Comment (osborneMs) [reply src/Morphir/IR/Repo.elm]: Fixed

File: src/Morphir/IR/Repo.elm (hunk 3 above)
Comment (osborneMs) [reply src/Morphir/IR/Repo.elm]: Fixed

**Review by AttilaMihaly — APPROVED** [review]

**Review by AttilaMihaly — APPROVED** [review]

**Pull request merged**